Sources/AudioKit/Internals/Utilities/Atomic.swift

```diff
@@ -0,0 +1,81 @@
+//  Atomic.swift
+//  MIDIKit
+//  Created by Steffan Andrews on 2020-12-20.
+
+import Darwin
+
+/// Defines a basic signature to which all locks will conform. Provides the basis for atomic access to stuff.
+private protocol ThreadLock {
+    init()
+
+    /// Lock a resource for writing. So only one thing can write, and nothing else can read or write.
+    func writeLock()
+
+    /// Lock a resource for reading. Other things can also lock for reading at the same time, but nothing else can write at that time.
+    func readLock()
+
+    /// Unlock a resource
+    func unlock()
+}
+
+private final class RWThreadLock: ThreadLock {
+    private var lock = pthread_rwlock_t()
+
+    init() {
+        guard pthread_rwlock_init(&lock, nil) == 0 else {
+            preconditionFailure("Unable to initialize the lock")
+        }
+    }
+
+    deinit {
+        pthread_rwlock_destroy(&lock)
+    }
+
+    func writeLock() {
+        pthread_rwlock_wrlock(&lock)
+    }
+
+    func readLock() {
+        pthread_rwlock_rdlock(&lock)
+    }
+
+    func unlock() {
+        pthread_rwlock_unlock(&lock)
+    }
+}
+
+/// A property wrapper that ensures atomic access to a value, meaning thread-safe with implicit serial read/write access.
+/// Multiple read accesses can potentially read at the same time, just not during a write.
+/// By using `pthread` to do the locking, this safer then using a `DispatchQueue/barrier` as there isn't a chance of priority inversion.
+@propertyWrapper
```

Comment:
what do you think about warning or erroring out if this property wrapper is used on the main thread?